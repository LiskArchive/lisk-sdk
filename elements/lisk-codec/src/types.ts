/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

export type BaseTypes = string | number | Buffer | bigint | boolean;

export interface GenericObject {
	[key: string]: GenericObject | BaseTypes | Array<BaseTypes | GenericObject>;
}

export interface SchemaPair {
	readonly [key: string]: SchemaProps;
}

export interface Schema {
	readonly $id: string;
	readonly type: string;
	readonly properties: Record<string, unknown>;
	readonly required?: string[];
}

export interface ValidatedSchema {
	readonly $id: string;
	readonly $schema: string;
	readonly type: string;
	readonly required?: string[];
	properties: SchemaPair;
}

export interface SchemaProps {
	readonly fieldNumber: number;
	readonly type?: string;
	readonly dataType?: string;
	readonly properties?: SchemaPair;
	readonly items?: SchemaObjectItem | SchemaScalarItem;
}

export interface SchemaObjectItem {
	readonly type: 'object';
	readonly fieldNumber: number;
	readonly properties: SchemaPair;
}

export interface SchemaScalarItem {
	readonly dataType: string;
	readonly type?: undefined;
}

export interface CompiledSchema {
	schemaProp: SchemaProps;
	propertyName: string;
	binaryKey: Buffer;
	dataPath: string[];
}

export type CompiledSchemasArray = Array<CompiledSchema | CompiledSchema[]>;

export interface CompiledSchemas {
	[key: string]: CompiledSchemasArray;
}

export interface Validator {
	addMetaSchema: (schema: Record<string, unknown>, key?: string) => void;
}
export interface IteratorReturnValue {
	// eslint-disable-next-line @typescript-eslint/no-explicit-any
	value: any;
	key: string;
}

export interface IteratableGenericObject extends GenericObject {
	// eslint-disable-next-line @typescript-eslint/no-explicit-any
	[Symbol.iterator](): Iterator<IteratorReturnValue>;
}
