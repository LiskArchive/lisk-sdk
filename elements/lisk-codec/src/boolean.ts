/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */
/* eslint-disable no-bitwise */
/* eslint-disable no-param-reassign */

export const writeBoolean = (value: boolean): Buffer =>
	value ? Buffer.from('01', 'hex') : Buffer.from('00', 'hex');

export const readBoolean = (buffer: Buffer, offset: number): [boolean, number] => {
	const val = buffer[offset];
	if (val !== 0x00 && val !== 0x01) {
		throw new Error('Invalid boolean bytes.');
	}
	return [val !== 0x00, 1];
};
