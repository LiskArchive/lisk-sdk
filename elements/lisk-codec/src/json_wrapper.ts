/* eslint-disable no-param-reassign */
/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */
import { address } from '@liskhq/lisk-cryptography';
import {
	BaseTypes,
	IteratableGenericObject,
	IteratorReturnValue,
	SchemaPair,
	SchemaProps,
} from './types';

type mapperFunction = (value: BaseTypes, format?: string) => BaseTypes;

interface mappersInterface {
	toJSON: {
		readonly [key: string]: mapperFunction;
	};
	fromJSON: {
		readonly [key: string]: mapperFunction;
	};
}

/* eslint-disable @typescript-eslint/explicit-function-return-type */
const mappers: mappersInterface = {
	toJSON: {
		uint32: value => value as number,
		sint32: value => value as number,
		uint64: value => (value as BigInt).toString(),
		sint64: value => (value as BigInt).toString(),
		string: value => value as string,
		bytes: (value, format?: string) => {
			if (!format || format === 'hex') {
				return (value as Buffer).toString('hex');
			}
			if (format === 'base64') {
				return (value as Buffer).toString('base64');
			}
			if (format === 'lisk32') {
				return address.getLisk32AddressFromAddress(value as Buffer);
			}
			throw new Error(`Unknown format ${format}.`);
		},
		boolean: value => value as boolean,
	},
	fromJSON: {
		uint32: value => value as number,
		sint32: value => value as number,
		uint64: value => BigInt(value),
		sint64: value => BigInt(value),
		string: value => value as string,
		bytes: (value, format?: string) => {
			if (!format || format === 'hex') {
				return Buffer.from(value as string, 'hex');
			}
			if (format === 'base64') {
				return Buffer.from(value as string, 'base64');
			}
			if (format === 'lisk32') {
				return address.getAddressFromLisk32Address(value as string);
			}
			throw new Error(`Unknown format ${format}.`);
		},
		boolean: value => value as boolean,
	},
};
/* eslint-enable @typescript-eslint/explicit-function-return-type */

const findObjectByPath = (message: SchemaProps, pathArr: string[]): SchemaProps | undefined => {
	let result: SchemaProps = message;
	for (let i = 0; i < pathArr.length; i += 1) {
		if (!result.properties && !result.items) {
			return undefined;
		}
		if (result.properties) {
			result = result.properties[pathArr[i]];
		} else if (result.items) {
			const x = (result.items as SchemaProps).properties as SchemaPair;
			result = x[pathArr[i]];
		}
	}
	return result;
};

const isObject = (item: unknown): boolean =>
	typeof item === 'object' && item !== null && !Array.isArray(item) && !Buffer.isBuffer(item);

export const iterator = function iterator(
	this: IteratableGenericObject,
): {
	next: () => {
		done: boolean;
		value: IteratorReturnValue;
	};
} {
	let index = 0;
	const properties = Object.keys(this);
	let Done = false;
	return {
		next: (): { done: boolean; value: IteratorReturnValue } => {
			Done = index >= properties.length;
			const obj = {
				done: Done,
				// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-member-access
				value: { value: this[properties[index]], key: properties[index] },
			};
			index += 1;
			return obj;
		},
	};
};

export const recursiveTypeCast = (
	mode: 'toJSON' | 'fromJSON',
	object: IteratableGenericObject,
	schema: SchemaProps,
	dataPath: string[],
): void => {
	for (const { key, value } of object) {
		if (isObject(value)) {
			dataPath.push(key);

			(value as IteratableGenericObject)[Symbol.iterator] = iterator;
			recursiveTypeCast(mode, value, schema, dataPath);
			dataPath.pop();

			delete (value as IteratableGenericObject)[(Symbol.iterator as unknown) as string];
		} else if (Array.isArray(value)) {
			dataPath.push(key);
			const schemaProp = findObjectByPath(schema, dataPath);
			if (schemaProp?.items?.type === 'object') {
				for (let i = 0; i < value.length; i += 1) {
					// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
					const arrayObject = value[i];

					(arrayObject as IteratableGenericObject)[Symbol.iterator] = iterator;
					recursiveTypeCast(mode, arrayObject, schema, dataPath);

					// eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
					delete arrayObject[Symbol.iterator];
				}
			} else {
				for (let i = 0; i < value.length; i += 1) {
					if (schemaProp === undefined || schemaProp.items === undefined) {
						delete object[key];
						dataPath.pop();
						continue;
					}

					// eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-explicit-any
					(object[key] as any)[i] = mappers[mode][schemaProp.items.dataType](
						value[i],
						schemaProp.items.format,
					);
				}
			}
			dataPath.pop();
		} else {
			dataPath.push(key);
			const schemaProp = findObjectByPath(schema, dataPath);

			if (schemaProp === undefined) {
				delete object[key];
				dataPath.pop();
				continue;
			}

			object[key] = mappers[mode][(schemaProp.dataType as unknown) as string](
				value,
				schemaProp.format,
			);

			delete object[(Symbol.iterator as unknown) as string];
			dataPath.pop();
		}
	}
	delete object[(Symbol.iterator as unknown) as string];
};
