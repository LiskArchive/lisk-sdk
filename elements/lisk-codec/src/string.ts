/* eslint-disable no-bitwise */
/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */
import { readBytes, writeBytes } from './bytes';

export const writeString = (value: string): Buffer => {
	const normalized = value.normalize('NFC');
	const stringBuffer = Buffer.from(normalized, 'utf8');

	return writeBytes(stringBuffer);
};

export const readString = (buffer: Buffer, offset: number): [string, number] => {
	const [value, size] = readBytes(buffer, offset);
	const decodedStr = new TextDecoder('utf8', { fatal: true }).decode(value);
	if (decodedStr.normalize('NFC') !== decodedStr) {
		throw new Error('UTF8 bytes include non-normalized bytes');
	}
	return [decodedStr, size];
};
