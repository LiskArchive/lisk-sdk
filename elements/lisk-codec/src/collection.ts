/* eslint-disable no-param-reassign */
/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */
import { GenericObject, CompiledSchema, CompiledSchemasArray } from './types';
import {
	writeSInt32,
	writeSInt64,
	writeUInt32,
	writeUInt64,
	readUInt32,
	readSInt32,
	readSInt64,
	readUInt64,
} from './varint';
import { writeString, readString } from './string';
import { writeBytes, readBytes } from './bytes';
import { writeBoolean, readBoolean } from './boolean';
import { readKey } from './keys';

const _readers: {
	// eslint-disable-next-line @typescript-eslint/no-explicit-any
	readonly [key: string]: (value: Buffer, offset: number) => any;
} = {
	uint32: readUInt32,
	sint32: readSInt32,
	uint64: readUInt64,
	sint64: readSInt64,
	string: readString,
	bytes: readBytes,
	boolean: readBoolean,
};

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const _writers: { readonly [key: string]: (value: any) => Buffer } = {
	uint32: writeUInt32,
	sint32: writeSInt32,
	uint64: writeUInt64,
	sint64: writeSInt64,
	string: writeString,
	bytes: writeBytes,
	boolean: writeBoolean,
};

export const writeObject = (
	compiledSchema: CompiledSchemasArray,
	message: GenericObject,
	chunks: Buffer[],
): [Buffer[], number] => {
	let simpleObjectSize = 0;
	for (let i = 0; i < compiledSchema.length; i += 1) {
		const property = compiledSchema[i];
		if (Array.isArray(property)) {
			const headerProp = property[0];
			if (headerProp.schemaProp.type === 'array') {
				// eslint-disable-next-line no-use-before-define
				const [, size] = writeArray(
					property,
					message[headerProp.propertyName] as Array<unknown>,
					chunks,
				);
				simpleObjectSize += size;
				continue;
			}
			// Write the key for container object
			chunks.push(headerProp.binaryKey);
			const [encodedValues, totalWrittenSize] = writeObject(
				property,
				message[headerProp.propertyName] as GenericObject,
				[],
			);
			// Add nested object size to total size
			const objectSize = _writers.uint32(totalWrittenSize);
			simpleObjectSize += objectSize.length + headerProp.binaryKey.length;
			chunks.push(objectSize);
			for (let e = 0; e < encodedValues.length; e += 1) {
				simpleObjectSize += encodedValues[e].length;
				chunks.push(encodedValues[e]);
			}
		} else {
			// This is the header object so it does not need to be written
			if (property.schemaProp.type === 'object') {
				continue;
			}
			const value = message[property.propertyName];
			// Missing properties are not encoded as per LIP-0027
			if (value === undefined) {
				continue;
			}

			const {
				schemaProp: { dataType },
				binaryKey,
			} = property;
			if (dataType === undefined) {
				throw new Error('Compiled Schema is corrupted as "dataType" can not be undefined.');
			}

			const binaryValue = _writers[dataType](value);

			chunks.push(binaryKey);
			chunks.push(binaryValue);
			simpleObjectSize += binaryKey.length + binaryValue.length;
		}
	}
	return [chunks, simpleObjectSize];
};

export const readObject = (
	message: Buffer,
	offset: number,
	compiledSchema: CompiledSchemasArray,
	terminateIndex: number,
): [GenericObject, number] => {
	let index = offset;
	const result: GenericObject = {};
	for (let i = 0; i < compiledSchema.length; i += 1) {
		const typeSchema = compiledSchema[i];
		if (Array.isArray(typeSchema)) {
			// Takeout the root wireType and field number
			if (typeSchema[0].schemaProp.type === 'array') {
				if (index >= terminateIndex) {
					result[typeSchema[0].propertyName] = [];
					continue;
				}
				// eslint-disable-next-line no-use-before-define
				const [arr, nextOffset] = readArray(message, index, typeSchema, terminateIndex);
				result[typeSchema[0].propertyName] = arr;
				index = nextOffset;
			} else if (typeSchema[0].schemaProp.type === 'object') {
				const [key, keySize] = readUInt32(message, index);
				const [fieldNumber] = readKey(key);
				// case where field number reading is not as expected from schema
				if (fieldNumber !== typeSchema[0].schemaProp.fieldNumber) {
					throw new Error('Invalid field number while decoding.');
				}

				index += keySize;
				// Takeout the length
				const [objectSize, objectSizeLength] = readUInt32(message, index);
				index += objectSizeLength;
				const [obj, nextOffset] = readObject(message, index, typeSchema, objectSize + index);
				result[typeSchema[0].propertyName] = obj;
				index = nextOffset;
			} else {
				throw new Error('Invalid container type.');
			}
			continue;
		}
		if (typeSchema.schemaProp.type === 'object' || typeSchema.schemaProp.type === 'array') {
			// typeSchema is header, and we ignore this
			continue;
		}
		// case where message length is shorter than what the schema expects
		if (message.length <= index) {
			throw new Error(
				`Message does not contain a property for fieldNumber: ${typeSchema.schemaProp.fieldNumber}.`,
			);
		}
		// Takeout the root wireType and field number
		const [key, keySize] = readUInt32(message, index);
		const [fieldNumber] = readKey(key);
		if (fieldNumber !== typeSchema.schemaProp.fieldNumber) {
			throw new Error('Invalid field number while decoding.');
		}
		// Index is only incremented when the key is actually used
		index += keySize;
		// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
		const [scalarValue, scalarSize] = _readers[typeSchema.schemaProp.dataType as string](
			message,
			index,
		);
		index += scalarSize;
		// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
		result[typeSchema.propertyName] = scalarValue;
	}

	// case where message length is longer than what the schema expects
	if (index !== terminateIndex) {
		throw new Error(
			`Invalid terminate index. Index is ${index} but terminateIndex is ${terminateIndex}`,
		);
	}

	return [result, index];
};

export const readArray = (
	message: Buffer,
	offset: number,
	compiledSchema: CompiledSchemasArray,
	terminateIndex: number,
	// eslint-disable-next-line @typescript-eslint/no-explicit-any
): [Array<any>, number] => {
	// Takeout the root wireType and field number
	let index = offset;
	if (index >= message.length) {
		return [[], index];
	}
	const startingByte = message[index];
	const [rootSchema, typeSchema] = compiledSchema;
	// Peek the current key, and if not the same fieldnumber, skip
	const [key] = readUInt32(message, index);
	const [fieldNumber] = readKey(key);
	if (fieldNumber !== (rootSchema as CompiledSchema).schemaProp.fieldNumber) {
		return [[], index];
	}
	// eslint-disable-next-line @typescript-eslint/no-explicit-any
	const result: Array<any> = [];
	// Case of object
	if (Array.isArray(typeSchema)) {
		// handle as object
		const [nestedTypeSchema] = typeSchema;
		if (nestedTypeSchema.schemaProp.type === 'object') {
			// If still the next bytes is the same key, it is still element of array
			// Also, in case of object, inside of array, it checks the size of the object
			while (message[index] === startingByte && index !== terminateIndex) {
				const [, wire2KeySize] = readUInt32(message, index);
				index += wire2KeySize;
				// Takeout the length
				const [objectSize, objectSizeLength] = readUInt32(message, index);
				// for object, length is not used
				index += objectSizeLength;
				if (objectSize === 0) {
					continue;
				}
				// If array of object, it also gives the terminating index of the particular object
				const terminatingObjectSize = index + objectSize;
				// readObject returns Next offset, not index used
				const [res, nextOffset] = readObject(message, index, typeSchema, terminatingObjectSize);
				result.push(res);
				index = nextOffset;
			}

			return [result, index];
		}
		throw new Error('Invalid container type');
	}
	// Case for string and bytes
	if (typeSchema.schemaProp.dataType === 'string' || typeSchema.schemaProp.dataType === 'bytes') {
		// If still the next bytes is the same key, it is still element of array
		// Also, in case of object inside of array, it checks the size of the object
		while (message[index] === startingByte && index !== terminateIndex) {
			const [, wire2KeySize] = readUInt32(message, index);
			index += wire2KeySize;
			// wireType2LengthSize is used while decoding string or bytes, therefore it's not subtracted unless it's zero
			const [wireType2Length, wireType2LengthSize] = readUInt32(message, index);
			if (wireType2Length === 0) {
				if (typeSchema.schemaProp.dataType === 'string') {
					result.push('');
				} else {
					result.push(Buffer.alloc(0));
				}
				index += wireType2LengthSize;
				// Add default value
				continue;
			}
			// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
			const [res, wire2Size] = _readers[typeSchema.schemaProp.dataType as string](message, index);
			result.push(res);
			index += wire2Size;
		}

		return [result, index];
	}
	const [, keySize] = readUInt32(message, index);
	index += keySize;
	// Takeout the length
	const [arrayLength, wireType2Size] = readUInt32(message, index);
	index += wireType2Size;
	// Case for varint and boolean
	const end = index + arrayLength;
	while (index < end) {
		// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
		const [res, size] = _readers[typeSchema.schemaProp.dataType as string](message, index);
		result.push(res);
		index += size;
	}

	return [result, index];
};

export const writeArray = (
	compiledSchema: CompiledSchema[],
	message: Array<unknown>,
	chunks: Buffer[],
): [Buffer[], number] => {
	if (message.length === 0) {
		return [chunks, 0];
	}
	let totalSize = 0;
	const [rootSchema, typeSchema] = compiledSchema;
	// Array of object
	if (Array.isArray(typeSchema)) {
		for (let i = 0; i < message.length; i += 1) {
			const [res, objectSize] = writeObject(typeSchema, message[i] as GenericObject, []);
			chunks.push(rootSchema.binaryKey);
			const size = _writers.uint32(objectSize);
			chunks.push(size);
			for (let j = 0; j < res.length; j += 1) {
				chunks.push(res[j]);
			}
			totalSize += objectSize + size.length + rootSchema.binaryKey.length;
		}
		return [chunks, totalSize];
	}
	// Array of string or bytes
	if (typeSchema.schemaProp.dataType === 'string' || typeSchema.schemaProp.dataType === 'bytes') {
		for (let i = 0; i < message.length; i += 1) {
			const res = _writers[typeSchema.schemaProp.dataType as string](message[i]);
			chunks.push(rootSchema.binaryKey);
			chunks.push(res);
			totalSize += res.length + rootSchema.binaryKey.length;
		}
		return [chunks, totalSize];
	}
	// Array of number or boolean
	chunks.push(rootSchema.binaryKey);
	// Insert size
	const contents = [];
	let contentSize = 0;
	for (let i = 0; i < message.length; i += 1) {
		const res = _writers[typeSchema.schemaProp.dataType as string](message[i]);
		contents.push(res);
		contentSize += res.length;
	}
	const arrayLength = _writers.uint32(contentSize);
	chunks.push(arrayLength);
	for (let i = 0; i < contents.length; i += 1) {
		chunks.push(contents[i]);
	}
	totalSize += rootSchema.binaryKey.length + contentSize + arrayLength.length;
	return [chunks, totalSize];
};
