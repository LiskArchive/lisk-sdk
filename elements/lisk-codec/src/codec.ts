/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

import { validator, liskSchemaIdentifier } from '@liskhq/lisk-validator';
import { objects as objectUtils } from '@liskhq/lisk-utils';
import { generateKey } from './utils';
import { readObject, writeObject } from './collection';

import {
	Schema,
	CompiledSchema,
	CompiledSchemas,
	CompiledSchemasArray,
	GenericObject,
	IteratableGenericObject,
	ValidatedSchema,
	SchemaProps,
} from './types';

import { iterator, recursiveTypeCast } from './json_wrapper';

export const validateSchema = (schema: {
	// eslint-disable-next-line
	[key: string]: any;
	$schema?: string;
	$id?: string;
}): boolean => {
	// TODO: https://github.com/ajv-validator/ajv/issues/1221
	//
	// Ajv compiles schema and cache it when either "validate" or "compile" called
	//
	// Due to issue mentioned above we have to compile the schema
	// manually our self. That requires to clear the cache manually so
	// any subsequent request to validate or compile may not fail.
	//
	// We have to clear cache once on start and once at the end
	// to cover both cases in which order "validate" or "compile"
	// is called from main code.

	validator.removeSchema(schema.$id);

	const schemaToValidate = {
		...schema,
		$schema: schema.$schema ?? liskSchemaIdentifier,
	};

	validator.validateSchema(schemaToValidate);

	try {
		// To validate keyword schema we have to compile it
		// Ajv `validateSchema` does not validate keyword meta schema
		// https://github.com/ajv-validator/ajv/issues/1221
		validator.compile(schemaToValidate);
	} finally {
		validator.removeSchema(schema.$id);
	}

	return true;
};

export class Codec {
	private _compileSchemas: CompiledSchemas = {};

	public addSchema(schema: Schema): boolean {
		validateSchema(schema);

		const schemaName = schema.$id;
		this._compileSchemas[schemaName] = this._compileSchema(schema as ValidatedSchema, [], []);

		return true;
	}

	public encode(schema: Schema, message: object): Buffer {
		if (this._compileSchemas[schema.$id] === undefined) {
			this.addSchema(schema);
		}

		const compiledSchema = this._compileSchemas[schema.$id];
		const res = writeObject(compiledSchema, message as GenericObject, []);
		return Buffer.concat(res[0]);
	}

	public decode<T>(schema: Schema, message: Buffer): T {
		if (this._compileSchemas[schema.$id] === undefined) {
			this.addSchema(schema);
		}
		const compiledSchema = this._compileSchemas[schema.$id];
		const [res] = readObject(message, 0, compiledSchema, message.length);

		return res as unknown as T;
	}

	// For performance applications use decode() instead!
	public decodeJSON<T>(schema: Schema, message: Buffer): T {
		const decodedMessage: IteratableGenericObject = this.decode(schema, message);

		const jsonMessageAsObject = this.toJSON(schema, decodedMessage);
		return jsonMessageAsObject as unknown as T;
	}

	// For performance applications use encode() instead!
	public encodeJSON(schema: Schema, message: object): Buffer {
		const objectFromJson = this.fromJSON(schema, message);
		return this.encode(schema, objectFromJson);
	}

	public toJSON<T = object>(schema: Schema, message: object): T {
		const messageCopy = objectUtils.cloneDeep(message);
		(messageCopy as IteratableGenericObject)[Symbol.iterator] = iterator;

		recursiveTypeCast(
			'toJSON',
			messageCopy as IteratableGenericObject,
			schema as unknown as SchemaProps,
			[],
		);
		return messageCopy as unknown as T;
	}

	public fromJSON<T = object>(schema: Schema, message: object): T {
		const messageCopy = objectUtils.cloneDeep(message);
		(messageCopy as IteratableGenericObject)[Symbol.iterator] = iterator;

		recursiveTypeCast(
			'fromJSON',
			messageCopy as IteratableGenericObject,
			schema as unknown as SchemaProps,
			[],
		);
		return messageCopy as unknown as T;
	}

	public clearCache(): void {
		this._compileSchemas = {};
	}

	private _compileSchema(
		schema: ValidatedSchema | SchemaProps,
		compiledSchema: CompiledSchemasArray,
		dataPath: string[],
	): CompiledSchemasArray {
		if (schema.type === 'object') {
			const { properties } = schema;
			if (properties === undefined) {
				throw new Error('Invalid schema. Missing "properties" property');
			}
			for (const property of Object.values(properties)) {
				if (!('fieldNumber' in property)) {
					throw new Error('Invalid schema. Missing "fieldNumber" in properties');
				}
			}
			const currentDepthSchema = Object.entries(properties).sort(
				(a, b) => a[1].fieldNumber - b[1].fieldNumber,
			);

			for (let i = 0; i < currentDepthSchema.length; i += 1) {
				const [schemaPropertyName, schemaPropertyValue] = currentDepthSchema[i];
				if (schemaPropertyValue.type === 'object') {
					if (!('fieldNumber' in schemaPropertyValue)) {
						throw new Error('Invalid schema. Missing "fieldNumber" in properties');
					}
					// Object recursive case
					dataPath.push(schemaPropertyName);
					const nestedSchema = [
						{
							propertyName: schemaPropertyName,
							schemaProp: {
								type: schemaPropertyValue.type,
								fieldNumber: schemaPropertyValue.fieldNumber,
							},
							dataPath: [...dataPath],
							binaryKey: generateKey(schemaPropertyValue),
						},
					];
					const res = this._compileSchema(schemaPropertyValue, nestedSchema, dataPath);
					compiledSchema.push(res as CompiledSchema[]);
					dataPath.pop();
				} else if (schemaPropertyValue.type === 'array') {
					// Array recursive case
					if (schemaPropertyValue.items === undefined) {
						throw new Error('Invalid schema. Missing "items" property for Array schema');
					}
					if (!('fieldNumber' in schemaPropertyValue)) {
						throw new Error('Invalid schema. Missing "fieldNumber" in properties');
					}
					dataPath.push(schemaPropertyName);
					if (schemaPropertyValue.items.type === 'object') {
						const nestedSchema = [
							{
								propertyName: schemaPropertyName,
								schemaProp: {
									type: 'object',
									fieldNumber: schemaPropertyValue.fieldNumber,
								},
								dataPath: [...dataPath],
								binaryKey: generateKey(schemaPropertyValue),
							},
						];
						const res = this._compileSchema(schemaPropertyValue.items, nestedSchema, dataPath);
						compiledSchema.push([
							{
								propertyName: schemaPropertyName,
								schemaProp: {
									type: schemaPropertyValue.type,
									fieldNumber: schemaPropertyValue.fieldNumber,
								},
								dataPath: [...dataPath],
								binaryKey: generateKey(schemaPropertyValue),
							},
							res as unknown as CompiledSchema,
						]);
						dataPath.pop();
					} else {
						compiledSchema.push([
							{
								propertyName: schemaPropertyName,
								schemaProp: {
									type: schemaPropertyValue.type,
									fieldNumber: schemaPropertyValue.fieldNumber,
								},
								dataPath: [...dataPath],
								binaryKey: generateKey(schemaPropertyValue),
							},
							{
								propertyName: schemaPropertyName,
								schemaProp: {
									dataType: schemaPropertyValue.items.dataType,
									fieldNumber: schemaPropertyValue.fieldNumber,
								},
								dataPath: [...dataPath],
								binaryKey: generateKey(schemaPropertyValue),
							},
						]);
						dataPath.pop();
					}
				} else {
					// Base case
					compiledSchema.push({
						propertyName: schemaPropertyName,
						schemaProp: schemaPropertyValue,
						dataPath: [...dataPath],
						binaryKey: generateKey(schemaPropertyValue),
					});
				}
			}
		}
		return compiledSchema;
	}
}

export const codec = new Codec();
