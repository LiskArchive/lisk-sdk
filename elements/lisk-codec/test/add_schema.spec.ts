/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

import { objects } from '@liskhq/lisk-utils';
import { testCases as objectsTestCases } from '../fixtures/objects_encodings.json';
import { codec } from '../src/codec';

describe('addSchema', () => {
	// Arrange
	const objectFixtureInput = objectsTestCases[0].input;

	it('should add schema and keep it in cache', () => {
		const object = {
			...objectFixtureInput.object,
			balance: BigInt(objectFixtureInput.object.balance as unknown as string),
			address: Buffer.from(objectFixtureInput.object.address as unknown as string, 'hex'),
		};

		codec.encode(objectFixtureInput.schema, object);

		expect((codec as any)._compileSchemas.object11).toMatchSnapshot();
	});

	it('should throw if schema does not have fieldNumber in properties at root level', () => {
		const { schema } = objectFixtureInput;
		const customSchema = objects.cloneDeep(schema);
		// Remove the field number in properties at root level
		delete (customSchema as any).properties.asset.fieldNumber;

		expect(() => codec.addSchema(customSchema)).toThrow(
			'Invalid schema. Missing "fieldNumber" in properties',
		);
	});

	it('should throw if schema does not have fieldNumber in properties at nested level 1', () => {
		const { schema } = objectFixtureInput;
		const customSchema = objects.cloneDeep(schema);
		// Remove the field number in properties at nested level 1
		delete (customSchema as any).properties.asset.properties.fooBar.fieldNumber;

		expect(() => codec.addSchema(customSchema)).toThrow(
			'Invalid schema. Missing "fieldNumber" in properties',
		);
	});

	it('should throw if schema does not have fieldNumber in properties at nested level 2', () => {
		const { schema } = objectFixtureInput;
		const customSchema = objects.cloneDeep(schema);
		// Remove the field number in properties at nested level 2
		delete (customSchema as any).properties.asset.properties.fooBar.properties.foo.fieldNumber;

		expect(() => codec.addSchema(customSchema)).toThrow(
			'Invalid schema. Missing "fieldNumber" in properties',
		);
	});
});
