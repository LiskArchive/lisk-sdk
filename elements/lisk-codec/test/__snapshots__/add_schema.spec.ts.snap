// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addSchema should add schema and keep it in cache 1`] = `undefined`;
